crates/api/src/config.rs

```diff
@@ -14,68 +15,236 @@ fn tc<S: serde::Serialize, D: serde::de::DeserializeOwned>(
     .map_err(|e| K2Error::other_src("decode", e))
 }
 
-/// Denotes a type used to configure a specific kitsune2 module.
-///
-/// Note, the types defined in this struct are specifically for configuration
-/// that cannot be changed at runtime, the likes of which might be found
-/// in a configuration file.
-///
-/// If a specific module has a config that can be changed at runtime, the
-/// component found in this type might be a `default_` prefixed version
-/// of it, then the runtime value can be altered through different means.
-///
-/// It is highly recommended that you expose this struct in your module
-/// docs to help devs using your module understand how to configure it.
-pub trait ModConfig:
-    'static
-    + Sized
-    + Default
-    + std::fmt::Debug
-    + serde::Serialize
-    + serde::de::DeserializeOwned
-    + Send
-    + Sync
-{
+/// A callback to be invoked if the config value is updated at runtime.
+pub type ConfigUpdateCb =
+    Arc<dyn Fn(serde_json::Value) + 'static + Send + Sync>;
+
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+#[serde(transparent, rename_all = "camelCase")]
+struct ConfigEntry {
+    pub value: serde_json::Value,
+    #[serde(skip, default)]
+    pub update_cb: Option<ConfigUpdateCb>,
+}
+
+impl std::fmt::Debug for ConfigEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.value.fmt(f)
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(untagged, rename_all = "camelCase")]
+enum ConfigMap {
+    ConfigMap(BTreeMap<String, Box<Self>>),
+    ConfigEntry(ConfigEntry),
+}
+
+impl Default for ConfigMap {
+    fn default() -> Self {
+        Self::ConfigMap(BTreeMap::new())
+    }
+}
+
+struct Inner {
+    map: ConfigMap,
+    are_defaults_set: bool,
+    is_runtime: bool,
```

Comment:
Does this mean that the runtime logic is module scoped? So "you can replace fetch config at runtime" but "you cannot replace space config at runtime"?

If I'm understanding that right, I think this belongs at the `ConfigEntry` level?

Reply:
Or maybe I'm misunderstanding this field because an update handler would have to be registered per config entry for it to be possible to update it anyway. So is this just about "some values in this map can be updated"?

Reply:
This is purely about providing a warning if a config is updated that doesn't have an update callback registered. So we mark this when we switch to runtime, and thereafter the warning traces are emitted.

Reply:
Got it, cool okay!